examples/pipeline.rs

```diff
@@ -0,0 +1,180 @@
+use std::cell::RefCell;
+use std::io::Result;
+use std::net::SocketAddr;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::rc::Rc;
+use std::time::Duration;
+
+use futures::io::IoSliceMut;
+use futures::task::{Context, Poll};
+use futures::{AsyncBufReadExt, AsyncWriteExt};
+use futures_lite::{AsyncBufRead, AsyncRead, AsyncWrite};
+
+use glommio::channels::local_channel;
+use glommio::channels::local_channel::{LocalReceiver, LocalSender};
+use glommio::net::{TcpListener, TcpStream};
+use glommio::task::JoinHandle;
+use glommio::timer::sleep;
+use glommio::{enclose, Local, LocalExecutor, Task};
+
+type Request = Rc<Vec<u8>>;
+type Response = Rc<Vec<u8>>;
+
+pub fn server(port: u16) -> SocketAddr {
+    let listener = Rc::new(TcpListener::bind(("0.0.0.0", port)).unwrap());
+
+    Local::local(enclose!((listener) async move {
+        while let Ok(stream) = listener.accept().await {
+            let stream = Rc::new(RefCell::new(stream));
+
+            Local::local(enclose!((stream) async move {
+                pipeline(stream).await
+            })).detach();
+        }
+    }))
+    .detach();
+
+    listener.local_addr().unwrap()
+}
+
+async fn pipeline(stream: Rc<RefCell<TcpStream>>) {
```

Comment:
Is the idea that messages are received sequentially at the TCP level but then processed in parallel and potentially out-of-order? This is a great example of where a semaphore can be useful to limit parallelism.

I think we can merge such structure in a `sink`-`source` paradigm to glommio.

In general terms, we can require a `sink` to implement an adaptor to connect to the `source`, which will be native to glommio. The example then can be massively simplified.

I think a nice example could even be writing the stream to a file, using one of our file data structures - but up to you.

Reply:
> I think we can merge such structure in a `sink`-`source` paradigm to glommio.

It's a good idea to have something like that in glommio! It reminds me `akka-stream`, which is funny to work with. I'll spend some time trying to make a prototype of this.